src/psyclone/generator.py

```diff
@@ -242,33 +247,61 @@ def generate(filename, api="", kernel_paths=None, script_name=None,
     ast, invoke_info = parse(filename, api=api, invoke_name="invoke",
                              kernel_paths=kernel_paths,
                              line_length=line_length)
-    if api != "gocean1.0":
+
+    # TODO issue #1618 remove temporary flag, associated logic and Alg class
+    # plus tests (and ast variable).
+    #
+    # Temporary flag to allow optional testing of new LFRic metadata
+    # implementation (mainly that the PSyIR works with algorithm-layer
+    # code) whilst keeping the original implementation as default
+    # until it is working.
+    lfric_testing = False
+
+    if api in API_WITHOUT_ALGORITHM or \
+       (api == "dynamo0.3" and not lfric_testing):
         psy = PSyFactory(api, distributed_memory=distributed_memory)\
             .create(invoke_info)
         if script_name is not None:
             handle_script(script_name, psy, "trans")
+        alg_gen = None
 
-    alg_gen = None
-
-    if api == "gocean1.0":
+    elif api == "gocean1.0" or (api == "dynamo0.3" and lfric_testing):
         # Create language-level PSyIR from the Algorithm file
         reader = FortranReader()
-        psyir = reader.psyir_from_file(filename)
+        if api == "dynamo0.3":
+            # avoid undeclared builtin errors in PSyIR by adding "use
+            # builtins"
+            fp2_tree = parse_fp2(filename)
+            check_fp2_tree(fp2_tree, filename)
+            add_builtins_use(fp2_tree)
+            psyir = reader.psyir_from_source(str(fp2_tree))
+        else:
+            psyir = reader.psyir_from_file(filename)
 
         # Raise to Algorithm PSyIR
-        alg_trans = AlgTrans()
+        if api == "gocean1.0":
+            alg_trans = AlgTrans()
+        else:  # api == "dynamo0.3"
+            alg_trans = LFRicAlgTrans()
         alg_trans.apply(psyir)
 
+        if not psyir.walk(AlgorithmInvokeCall):
+            raise NoInvokesError(
```

Comment:
Is this going to change the current behaviour where we permit source files without invoke()s?